src/OpenFeatureSDK/OpenFeature.cs

```diff
@@ -11,7 +11,7 @@ namespace OpenFeatureSDK
     /// <seealso href="https://github.com/open-feature/spec/blob/main/specification/flag-evaluation.md#flag-evaluation-api"/>
     public sealed class OpenFeature
     {
-        private EvaluationContext _evaluationContext = new EvaluationContext();
+        private EvaluationContext _evaluationContext = EvaluationContext.Empty;
         private FeatureProvider _featureProvider = new NoOpFeatureProvider();
         private readonly List<Hook> _hooks = new List<Hook>();

```

Comment:
While we're addressing thread safety, do you think we need to worry about the hook list here being mutated in multiple threads?

Reply:
Yeah, that is true. I can take a look at that as well. Likely, to finish up thread safety, we may need some changes more than just the immutable builder.

Reply:
Yeah, the hooks are a problem.

Currently I think there is a problem with the global context as well. A threading issue, but also I think right now your client will have a global context based on when you get the client. So if you update the global context it doesn't affect clients you have already gotten from the SDK.

I suspect that is supposed to be like hooks, where adding a hook affect subsequent evaluations from existing clients.

Reply:
> I think right now your client will have a global context based on when you get the client.

I don't think this is explicitly spelled out in the spec, but it's implied. I think I should add something about it. It doesn't necessarily need to be addressed in this PR.

The same should be true with regards to the configured provider - the currently registered one should be used for all clients, regardless of when they were created.

cc @beeme1mr

Reply:
Another question, are hooks supposed to be executed in some specific order based on the order they were registered? There is an order of the sources of hooks, but multiple hooks registered at the API level for instance.

It affects the type of concurrency I would provide here. The C# `ConcurrentStack` has all the correct functionality, but the contents are reversed. (I can enumerable reverse this, so the cost shouldn't be that much.)
 
`ConcurrentQueue` puts things in the right order, but doesn't have a clear method. So you cannot atomically remove everything.

Or just making a synchronized collection, I prefer the concurrent solutions as they are generally lockless.

Reply:
Personally I didn't want to provide guarantees of hook ordering _within_ a stage. The spec never dictates any behavior there.

> ConcurrentQueue puts things in the right order, but doesn't have a clear method. So you cannot atomically remove everything.

Removing all hooks is also not specified. Most implementations do it b/c it helps with testing, but you could make this a package private method if you want. I don't think there's a real compelling use case for allowing authors to remove hooks, though I could be wrong about that.